// @tamagui/themes/dist/cjs/v3.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf, __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: !0 }) : target,
  mod
)), __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var v3_exports = {};
__export(v3_exports, {
  blue: () => import_colors.blue,
  blueDark: () => import_colors.blueDark,
  gray: () => import_colors8.gray,
  grayDark: () => import_colors8.grayDark,
  green: () => import_colors2.green,
  greenDark: () => import_colors2.greenDark,
  orange: () => import_colors4.orange,
  orangeDark: () => import_colors4.orangeDark,
  pink: () => import_colors6.pink,
  pinkDark: () => import_colors6.pinkDark,
  purple: () => import_colors7.purple,
  purpleDark: () => import_colors7.purpleDark,
  red: () => import_colors3.red,
  redDark: () => import_colors3.redDark,
  themes: () => themes,
  tokens: () => import_v3_themes.tokens,
  yellow: () => import_colors5.yellow,
  yellowDark: () => import_colors5.yellowDark
});
module.exports = __toCommonJS(v3_exports);
var generatedThemes = __toESM(require("./generated-v3")), import_v3_themes = require("./v3-themes"), import_colors = require("@tamagui/colors"), import_colors2 = require("@tamagui/colors"), import_colors3 = require("@tamagui/colors"), import_colors4 = require("@tamagui/colors"), import_colors5 = require("@tamagui/colors"), import_colors6 = require("@tamagui/colors"), import_colors7 = require("@tamagui/colors"), import_colors8 = require("@tamagui/colors");
const themes = generatedThemes;
