// @tamagui/react-native-web-internals/dist/cjs/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf,
  __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  },
  __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default")),
  __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
    value: mod,
    enumerable: !0
  }) : target, mod)),
  __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
    value: !0
  }), mod);
var index_exports = {};
__export(index_exports, {
  AccessibilityUtil: () => import_AccessibilityUtil.AccessibilityUtil,
  ImageLoader: () => import_ImageLoader.default,
  InteractionManager: () => import_InteractionManager.InteractionManager,
  Platform: () => import_Platform.default,
  StyleSheet: () => import_StyleSheet2.default,
  TextAncestorContext: () => import_TextAncestorContext.TextAncestorContext,
  TextInputState: () => import_TextInputState.default,
  UIManager: () => import_UIManager.default,
  canUseDOM: () => import_canUseDOM.default,
  colorProps: () => import_colorProps.colorProps,
  createDOMProps: () => import_createDOMProps.default,
  createEventHandle: () => import_createEventHandle.default,
  createReactDOMStyle: () => import_createReactDOMStyle.default,
  createSheet: () => import_dom.createSheet,
  dismissKeyboard: () => import_dismissKeyboard.default,
  flattenStyle: () => import_StyleSheet.flatten,
  getBoundingClientRect: () => import_getBoundingClientRect.default,
  isSelectionValid: () => import_isSelectionValid.default,
  isWebColor: () => import_isWebColor.default,
  multiplyStyleLengthValue: () => import_multiplyStyleLengthValue.default,
  normalizeColor: () => import_normalizeColor.default,
  pick: () => import_pick.default,
  processColor: () => import_processColor.processColor,
  requestIdleCallback: () => import_requestIdleCallback.default,
  setValueForStyles: () => import_setValueForStyles.default,
  stylesFromProps: () => import_createDOMProps2.stylesFromProps,
  unitlessNumbers: () => import_unitlessNumbers.default,
  useElementLayout: () => import_useElementLayout.default,
  useEvent: () => import_useEvent.default,
  useHover: () => import_useHover.default,
  useLayoutEffect: () => import_useLayoutEffect.default,
  useMergeRefs: () => import_useMergeRefs.useMergeRefs,
  usePlatformMethods: () => import_usePlatformMethods.usePlatformMethods,
  useStable: () => import_useStable.default
});
module.exports = __toCommonJS(index_exports);
__reExport(index_exports, require("./modules/AssetRegistry/index.cjs"), module.exports);
__reExport(index_exports, require("./modules/forwardedProps/index.cjs"), module.exports);
__reExport(index_exports, require("./modules/mergeRefs/index.cjs"), module.exports);
__reExport(index_exports, require("./modules/modality/index.cjs"), module.exports);
__reExport(index_exports, require("./modules/useLocale/index.cjs"), module.exports);
var import_usePlatformMethods = require("./modules/usePlatformMethods/index.cjs"),
  import_TextAncestorContext = require("./TextAncestorContext.cjs");
__reExport(index_exports, require("@tamagui/react-native-use-pressable"), module.exports);
__reExport(index_exports, require("@tamagui/react-native-use-responder-events"), module.exports);
var import_colorProps = require("./colorProps.cjs"),
  import_AccessibilityUtil = require("./modules/AccessibilityUtil/index.cjs"),
  import_canUseDOM = __toESM(require("./modules/canUseDOM.cjs")),
  import_createDOMProps = __toESM(require("./modules/createDOMProps/index.cjs")),
  import_createDOMProps2 = require("./modules/createDOMProps/index.cjs"),
  import_createReactDOMStyle = __toESM(require("./StyleSheet/compiler/createReactDOMStyle.cjs")),
  import_createEventHandle = __toESM(require("./modules/createEventHandle/index.cjs")),
  import_dismissKeyboard = __toESM(require("./modules/dismissKeyboard/index.cjs")),
  import_getBoundingClientRect = __toESM(require("./modules/getBoundingClientRect/index.cjs")),
  import_ImageLoader = __toESM(require("./modules/ImageLoader/index.cjs")),
  import_isSelectionValid = __toESM(require("./modules/isSelectionValid/index.cjs")),
  import_isWebColor = __toESM(require("./modules/isWebColor/index.cjs")),
  import_multiplyStyleLengthValue = __toESM(require("./modules/multiplyStyleLengthValue/index.cjs")),
  import_normalizeColor = __toESM(require("./modules/normalizeColor/index.cjs")),
  import_pick = __toESM(require("./modules/pick/index.cjs")),
  import_Platform = __toESM(require("./modules/Platform/index.cjs"));
__reExport(index_exports, require("./StyleSheet/preprocess.cjs"), module.exports);
var import_StyleSheet = require("./StyleSheet/index.cjs"),
  import_dom = require("./StyleSheet/dom/index.cjs"),
  import_requestIdleCallback = __toESM(require("./modules/requestIdleCallback/index.cjs")),
  import_setValueForStyles = __toESM(require("./modules/setValueForStyles/index.cjs")),
  import_TextInputState = __toESM(require("./modules/TextInputState/index.cjs")),
  import_UIManager = __toESM(require("./modules/UIManager/index.cjs")),
  import_unitlessNumbers = __toESM(require("./modules/unitlessNumbers/index.cjs")),
  import_useElementLayout = __toESM(require("./modules/useElementLayout/index.cjs")),
  import_useEvent = __toESM(require("./modules/useEvent/index.cjs")),
  import_useHover = __toESM(require("./modules/useHover/index.cjs")),
  import_useLayoutEffect = __toESM(require("./modules/useLayoutEffect/index.cjs")),
  import_useStable = __toESM(require("./modules/useStable/index.cjs")),
  import_InteractionManager = require("./modules/InteractionManager.cjs");
__reExport(index_exports, require("./modules/invariant.cjs"), module.exports);
var import_processColor = require("./modules/processColor/index.cjs"),
  import_StyleSheet2 = __toESM(require("./StyleSheet/index.cjs")),
  import_useMergeRefs = require("./modules/useMergeRefs/index.cjs");
