// @tamagui/react-native-web-lite/dist/esm/AppRegistry/AppContainer.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from "react";
import { StyleSheet } from "@tamagui/react-native-web-internals";
import View from "../View/index";
var RootTagContext = /* @__PURE__ */ React.createContext(null), AppContainer = /* @__PURE__ */ React.forwardRef(function(props, forwardedRef) {
  var { children, WrapperComponent } = props, innerView = /* @__PURE__ */ _jsx(View, {
    pointerEvents: "box-none",
    style: styles.appContainer,
    children
  }, 1);
  return WrapperComponent && (innerView = /* @__PURE__ */ _jsx(WrapperComponent, {
    children: innerView
  })), /* @__PURE__ */ _jsx(RootTagContext.Provider, {
    value: props.rootTag,
    children: /* @__PURE__ */ _jsx(View, {
      pointerEvents: "box-none",
      ref: forwardedRef,
      style: styles.appContainer,
      children: innerView
    })
  });
});
AppContainer.displayName = "AppContainer";
var AppContainer_default = AppContainer, styles = StyleSheet.create({
  appContainer: {
    flex: 1
  }
});
export {
  AppContainer_default as default
};
