// @tamagui/react-native-web-internals/dist/esm/modules/AccessibilityUtil/propsToAccessibilityComponent.native.js
import propsToAriaRole from "./propsToAriaRole";
var roleComponents = {
  article: "article",
  banner: "header",
  blockquote: "blockquote",
  code: "code",
  complementary: "aside",
  contentinfo: "footer",
  deletion: "del",
  emphasis: "em",
  figure: "figure",
  insertion: "ins",
  form: "form",
  list: "ul",
  listitem: "li",
  main: "main",
  navigation: "nav",
  region: "section",
  strong: "strong"
}, emptyObject = {}, propsToAccessibilityComponent = function() {
  var props = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : emptyObject;
  if (props.accessibilityRole === "label")
    return "label";
  var role = propsToAriaRole(props);
  if (role) {
    if (role === "heading") {
      var level = props.accessibilityLevel || props["aria-level"];
      return level != null ? `h${level}` : "h1";
    }
    return roleComponents[role];
  }
}, propsToAccessibilityComponent_default = propsToAccessibilityComponent;
export {
  propsToAccessibilityComponent_default as default
};
