// @tamagui/react-native-web-lite/dist/esm/DeviceInfo/index.native.js
import { canUseDOM } from "@tamagui/react-native-web-internals";
import Dimensions from "../Dimensions/index";
var DeviceInfo = {
  Dimensions: {
    get windowPhysicalPixels() {
      var { width, height, fontScale, scale } = Dimensions.get("window");
      return {
        width: width * scale,
        height: height * scale,
        scale,
        fontScale
      };
    },
    get screenPhysicalPixels() {
      var { width: width1, height: height1, fontScale: fontScale1, scale: scale1 } = Dimensions.get("screen");
      return {
        width: width1 * scale1,
        height: height1 * scale1,
        scale: scale1,
        fontScale: fontScale1
      };
    }
  },
  get locale() {
    if (canUseDOM)
      return navigator.languages ? navigator.languages[0] : navigator.language;
  },
  get totalMemory() {
    return canUseDOM ? navigator.deviceMemory : void 0;
  },
  get userAgent() {
    return canUseDOM ? navigator.userAgent : "";
  }
}, DeviceInfo_default = DeviceInfo;
export {
  DeviceInfo_default as default
};
