// @tamagui/static/dist/registerRequire.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var registerRequire_exports = {};
__export(registerRequire_exports, {
  getNameToPaths: () => getNameToPaths,
  registerRequire: () => registerRequire,
  setRequireResult: () => setRequireResult
});
module.exports = __toCommonJS(registerRequire_exports);
var import_node = require("esbuild-register/dist/node"), import_bundle = require("./extractor/bundle"), import_requireTamaguiCore = require("./helpers/requireTamaguiCore"), _process_env_TAMAGUI_IGNORE_BUNDLE_ERRORS, nameToPaths = {}, getNameToPaths = function() {
  return nameToPaths;
}, Module = require("node:module"), proxyWorm = require("@tamagui/proxy-worm"), isRegistered = !1, og, whitelisted = {
  react: !0
}, compiled = {};
function setRequireResult(name, result) {
  compiled[name] = result;
}
function registerRequire(platform) {
  var { proxyWormImports } = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {
    proxyWormImports: !1
  };
  if (isRegistered)
    return {
      tamaguiRequire: require,
      unregister: function() {
      }
    };
  var { unregister } = (0, import_node.register)({
    hookIgnoreNodeModules: !1
  });
  og || (og = Module.prototype.require), isRegistered = !0, Module.prototype.require = tamaguiRequire;
  function tamaguiRequire(path) {
    var _this = this;
    if (path === "tamagui" && platform === "native")
      return og.apply(this, [
        "tamagui/native"
      ]);
    if (path === "@tamagui/core" || path === "@tamagui/web")
      return (0, import_requireTamaguiCore.requireTamaguiCore)(platform, function(path2) {
        return og.apply(_this, [
          path2
        ]);
      });
    if (path in compiled)
      return compiled[path];
    if (import_bundle.esbuildIgnoreFilesRegex.test(path))
      return {};
    if (path === "react-native-svg")
      return og.apply(this, [
        "@tamagui/react-native-svg"
      ]);
    if (path === "@gorhom/bottom-sheet" || path.startsWith("react-native-reanimated") || path === "expo-linear-gradient" || path === "@expo/vector-icons" || path === "tamagui/linear-gradient")
      return proxyWorm;
    if (path === "react-native/package.json")
      return require("react-native-web/package.json");
    if (path === "@tamagui/react-native-web-lite" || path === "react-native" || path.startsWith("react-native/"))
      try {
        return require("react-native");
      } catch {
        return require("@tamagui/react-native-web-lite");
      }
    if (path in knownIgnorableModules)
      return proxyWorm;
    if (!whitelisted[path] && proxyWormImports && !path.includes(".tamagui-dynamic-eval"))
      return path === "tamagui" ? og.apply(this, [
        path
      ]) : proxyWorm;
    try {
      var out = og.apply(this, arguments);
      return out;
    } catch (err) {
      return !process.env.TAMAGUI_ENABLE_WARN_DYNAMIC_LOAD && path.includes("tamagui-dynamic-eval") ? void 0 : (allowedIgnores[path] || IGNORES === "true" || (!process.env.TAMAGUI_SHOW_FULL_BUNDLE_ERRORS && !process.env.DEBUG ? hasWarnedForModules.has(path) || (hasWarnedForModules.add(path), console.info(`  tamagui: skipping ${path} tamagui.dev/docs/intro/errors#warning-001`)) : console.error(`Tamagui failed loading "${path}"
  
  ${err.message}
  ${err.stack}

  `)), proxyWorm);
    }
  }
  return {
    tamaguiRequire,
    unregister: function() {
      unregister(), isRegistered = !1, Module.prototype.require = og;
    }
  };
}
var IGNORES = process.env.TAMAGUI_IGNORE_BUNDLE_ERRORS, extraIgnores = IGNORES === "true" ? [] : (_process_env_TAMAGUI_IGNORE_BUNDLE_ERRORS = process.env.TAMAGUI_IGNORE_BUNDLE_ERRORS) === null || _process_env_TAMAGUI_IGNORE_BUNDLE_ERRORS === void 0 ? void 0 : _process_env_TAMAGUI_IGNORE_BUNDLE_ERRORS.split(","), knownIgnorableModules = {
  "expo-modules": !0,
  solito: !0,
  ...Object.fromEntries((extraIgnores == null ? void 0 : extraIgnores.map(function(k) {
    return [
      k,
      !0
    ];
  })) || [])
}, hasWarnedForModules = /* @__PURE__ */ new Set(), allowedIgnores = {
  "expo-constants": !0,
  "./ExpoHaptics": !0,
  "./js/MaskedView": !0
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getNameToPaths,
  registerRequire,
  setRequireResult
});
