// @tamagui/react-native-web-lite/dist/esm/index.mjs
import { default as default2 } from "./createElement/index.mjs";
import { processColor, Platform, UIManager, canUseDOM, dismissKeyboard, isWebColor, mergeRefs, useMergeRefs, useEvent, useHover, ImageLoader, AccessibilityUtil, InteractionManager, LocaleProvider, createDOMProps, normalizeColor, usePlatformMethods, useLayoutEffect, processStyle, StyleSheet, clickProps, TextAncestorContext, flattenStyle } from "@tamagui/react-native-web-internals";
import { default as default3 } from "./render/index.mjs";
import { unmountComponentAtNode } from "react-dom";
import { default as default4 } from "./NativeModules/index.mjs";
import { findNodeHandle } from "./findNodeHandle.mjs";
import { default as default5 } from "./vendor/react-native/Animated/Easing.mjs";
import { default as default6 } from "./vendor/react-native/Animated/Animated.mjs";
import { default as default7 } from "./vendor/react-native/NativeEventEmitter/index.mjs";
import { default as default8 } from "./AccessibilityInfo/index.mjs";
import { default as default9 } from "./Alert/index.mjs";
import { default as default10 } from "./Appearance/index.mjs";
import { default as default11 } from "./AppRegistry/index.mjs";
import { default as default12 } from "./AppState/index.mjs";
import { default as default13 } from "./BackHandler/index.mjs";
import { default as default14 } from "./Clipboard/index.mjs";
import { default as default15 } from "./DeviceInfo/index.mjs";
import { default as default16 } from "./DeviceEmitter.mjs";
import { default as default17 } from "./DeviceEmitter.mjs";
import { default as default18 } from "./Dimensions/index.mjs";
import { default as default19 } from "./I18nManager/index.mjs";
import { default as default20 } from "./Keyboard/index.mjs";
import { default as default21 } from "./Linking/index.mjs";
import { default as default22 } from "./PanResponder/index.mjs";
import { default as default23 } from "./PixelRatio/index.mjs";
import { default as default24 } from "./Share/index.mjs";
import { default as default25 } from "./Vibration/index.mjs";
import { default as default26 } from "./UnimplementedView.mjs";
import { default as default27 } from "./UnimplementedView.mjs";
import { default as default28 } from "./vendor/react-native/VirtualizedList";
import { default as default29 } from "./FlatList.mjs";
import { default as default30 } from "./UnimplementedView.mjs";
import { default as default31 } from "./UnimplementedView.mjs";
import { default as default32 } from "./UnimplementedView.mjs";
import { default as default33 } from "./TouchableOpacity.mjs";
import { default as default34 } from "./TouchableOpacity.mjs";
import { default as default35 } from "./TouchableWithoutFeedback.mjs";
import { default as default36 } from "./ActivityIndicator/index.mjs";
import { default as default37 } from "./Image/index.mjs";
import { default as default38 } from "./ImageBackground/index.mjs";
import { default as default39 } from "./KeyboardAvoidingView/index.mjs";
import { default as default40 } from "./Modal/index.mjs";
import { default as default41 } from "./Pressable/index.mjs";
import { default as default42 } from "./RefreshControl/index.mjs";
import { default as default43 } from "./SafeAreaView/index.mjs";
import { default as default44 } from "./ScrollView/index.mjs";
import { default as default45 } from "./StatusBar/index.mjs";
import { default as default46 } from "./Text/index.mjs";
import { default as default47 } from "./TextInput/index.mjs";
import { default as default48 } from "./View/index.mjs";
import { default as default49 } from "./LogBox/index.mjs";
import { default as default50 } from "./useColorScheme/index.mjs";
import { useLocaleContext } from "./useLocaleContext/index.mjs";
import { default as default51 } from "./useWindowDimensions/index.mjs";
function requireNativeComponent(name) {
  return function () {
    return null;
  };
}
export { default8 as AccessibilityInfo, AccessibilityUtil, default36 as ActivityIndicator, default9 as Alert, default6 as Animated, default11 as AppRegistry, default12 as AppState, default10 as Appearance, default13 as BackHandler, default14 as Clipboard, default16 as DeviceEmitter, default17 as DeviceEventEmitter, default15 as DeviceInfo, default18 as Dimensions, default26 as DrawerLayoutAndroid, default5 as Easing, default29 as FlatList, default19 as I18nManager, default37 as Image, default38 as ImageBackground, ImageLoader, InteractionManager, default20 as Keyboard, default39 as KeyboardAvoidingView, default21 as Linking, LocaleProvider, default49 as LogBox, default40 as Modal, default7 as NativeEventEmitter, default4 as NativeModules, default22 as PanResponder, default23 as PixelRatio, Platform, default41 as Pressable, default42 as RefreshControl, default43 as SafeAreaView, default44 as ScrollView, default32 as SectionList, default24 as Share, default45 as StatusBar, StyleSheet, default27 as Switch, default46 as Text, TextAncestorContext, default47 as TextInput, default33 as Touchable, default30 as TouchableHighlight, default31 as TouchableNativeFeedback, default34 as TouchableOpacity, default35 as TouchableWithoutFeedback, UIManager, default25 as Vibration, default48 as View, default28 as VirtualizedList, canUseDOM, clickProps, createDOMProps, dismissKeyboard, findNodeHandle, flattenStyle, isWebColor, mergeRefs, normalizeColor, processColor, processStyle, default3 as render, requireNativeComponent, unmountComponentAtNode, default2 as unstable_createElement, default50 as useColorScheme, useEvent, useHover, useLayoutEffect, useLocaleContext, useMergeRefs, usePlatformMethods, default51 as useWindowDimensions };
