// @tamagui/babel-plugin-fully-specified/dist/esm/index.native.js
import { existsSync, lstatSync } from "node:fs";
import { dirname, extname, resolve } from "node:path";
var DEFAULT_OPTIONS = {
  ensureFileExists: !0,
  esExtensionDefault: ".mjs",
  tryExtensions: [
    ".js"
  ],
  esExtensions: [
    ".mjs"
  ]
};
function FullySpecified(api, rawOptions) {
  api.assertVersion(7);
  var options = {
    ...DEFAULT_OPTIONS,
    ...rawOptions
  }, importDeclarationVisitor = function(path, state) {
    var filePath = state.file.opts.filename;
    if (filePath) {
      var { node } = path;
      if (node.importKind !== "type") {
        var originalModuleSpecifier = node.source.value, fullySpecifiedModuleSpecifier = getFullySpecifiedModuleSpecifier(originalModuleSpecifier, {
          filePath,
          options
        });
        fullySpecifiedModuleSpecifier && (node.source.value = fullySpecifiedModuleSpecifier);
      }
    }
  }, exportDeclarationVisitor = function(path, state) {
    var filePath = state.file.opts.filename;
    if (filePath) {
      var { node } = path;
      if (node.exportKind !== "type") {
        var source = node.source;
        if (source) {
          var originalModuleSpecifier = source.value, fullySpecifiedModuleSpecifier = getFullySpecifiedModuleSpecifier(originalModuleSpecifier, {
            filePath,
            options
          });
          fullySpecifiedModuleSpecifier && (source.value = fullySpecifiedModuleSpecifier);
        }
      }
    }
  }, importVisitor = function(path, state) {
    var filePath = state.file.opts.filename;
    if (filePath) {
      var parent = path.parent;
      if (parent.type === "CallExpression") {
        var firstArgOfImportCall = parent.arguments[0];
        if (firstArgOfImportCall.type === "StringLiteral") {
          var originalModuleSpecifier = firstArgOfImportCall.value, fullySpecifiedModuleSpecifier = getFullySpecifiedModuleSpecifier(originalModuleSpecifier, {
            filePath,
            options
          });
          fullySpecifiedModuleSpecifier && (firstArgOfImportCall.value = fullySpecifiedModuleSpecifier);
        }
      }
    }
  };
  return {
    name: "babel-plugin-fully-specified",
    visitor: {
      ImportDeclaration: importDeclarationVisitor,
      ExportNamedDeclaration: exportDeclarationVisitor,
      ExportAllDeclaration: exportDeclarationVisitor,
      Import: importVisitor
    }
  };
}
function getFullySpecifiedModuleSpecifier(originalModuleSpecifier, param) {
  var { filePath, options } = param, fileExt = extname(filePath), fileDir = dirname(filePath), isDirectory = isLocalDirectory(resolve(fileDir, originalModuleSpecifier)), currentModuleExtension = extname(originalModuleSpecifier), { tryExtensions, esExtensions, esExtensionDefault, ensureFileExists } = options, targetModule = evaluateTargetModule({
    moduleSpecifier: originalModuleSpecifier,
    filenameDirectory: fileDir,
    filenameExtension: fileExt,
    currentModuleExtension,
    isDirectory,
    tryExtensions,
    esExtensions,
    esExtensionDefault,
    ensureFileExists
  });
  return targetModule === !1 ? null : targetModule;
}
function isLocalDirectory(absoluteDirectory) {
  return existsSync(absoluteDirectory) && lstatSync(absoluteDirectory).isDirectory();
}
function evaluateTargetModule(param) {
  var { moduleSpecifier, currentModuleExtension, isDirectory, filenameDirectory, filenameExtension, tryExtensions, esExtensions, esExtensionDefault, ensureFileExists } = param;
  if (currentModuleExtension && !esExtensions.includes(currentModuleExtension))
    return !1;
  var targetFile = resolve(filenameDirectory, moduleSpecifier);
  if (ensureFileExists) {
    var _iteratorNormalCompletion = !0, _didIteratorError = !1, _iteratorError = void 0;
    try {
      for (var _iterator = tryExtensions[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = !0) {
        var extension = _step.value;
        if (existsSync(targetFile + extension))
          return moduleSpecifier + esExtensionDefault;
      }
    } catch (err) {
      _didIteratorError = !0, _iteratorError = err;
    } finally {
      try {
        !_iteratorNormalCompletion && _iterator.return != null && _iterator.return();
      } finally {
        if (_didIteratorError)
          throw _iteratorError;
      }
    }
    isDirectory && !existsSync(resolve(filenameDirectory, currentModuleExtension ? moduleSpecifier : moduleSpecifier + esExtensionDefault)) && (moduleSpecifier = `${moduleSpecifier}/index`);
  } else return esExtensions.includes(filenameExtension), moduleSpecifier + esExtensionDefault;
  return !1;
}
export {
  FullySpecified as default
};
