// @tamagui/themes/dist/esm/v4-defaultTemplates.mjs
import { objectFromEntries, objectKeys } from "./helpers.mjs";
const lightShadowColor = "rgba(0,0,0,0.04)",
  lightShadowColorStrong = "rgba(0,0,0,0.085)",
  darkShadowColor = "rgba(0,0,0,0.2)",
  darkShadowColorStrong = "rgba(0,0,0,0.3)",
  shadows = {
    light: {
      shadowColor: lightShadowColorStrong,
      shadowColorHover: lightShadowColorStrong,
      shadowColorPress: lightShadowColor,
      shadowColorFocus: lightShadowColor
    },
    dark: {
      shadowColor: darkShadowColorStrong,
      shadowColorHover: darkShadowColorStrong,
      shadowColorPress: darkShadowColor,
      shadowColorFocus: darkShadowColor
    }
  },
  getTemplates = () => {
    const lightTemplates = getBaseTemplates("light"),
      darkTemplates = getBaseTemplates("dark");
    return {
      ...objectFromEntries(objectKeys(lightTemplates).map(name => [`light_${name}`, lightTemplates[name]])),
      ...objectFromEntries(objectKeys(darkTemplates).map(name => [`dark_${name}`, darkTemplates[name]]))
    };
  },
  getBaseTemplates = scheme => {
    const isLight = scheme === "light",
      bgIndex = 5,
      lighten = isLight ? -1 : 1,
      darken = -lighten,
      borderColor = bgIndex + 3,
      base = {
        accentBackground: 0,
        accentColor: -0,
        ...(isLight ? shadows.light : shadows.dark),
        background0: 1,
        background025: 2,
        background05: 3,
        background075: 4,
        color1: bgIndex,
        color2: bgIndex + 1,
        color3: bgIndex + 2,
        color4: bgIndex + 3,
        color5: bgIndex + 4,
        color6: bgIndex + 5,
        color7: bgIndex + 6,
        color8: bgIndex + 7,
        color9: bgIndex + 8,
        color10: bgIndex + 9,
        color11: bgIndex + 10,
        color12: bgIndex + 11,
        color0: -1,
        color025: -2,
        color05: -3,
        color075: -4,
        // the background, color, etc keys here work like generics - they make it so you
        // can publish components for others to use without mandating a specific color scale
        // the @tamagui/button Button component looks for `$background`, so you set the
        // dark_red_Button theme to have a stronger background than the dark_red theme.
        background: bgIndex,
        backgroundHover: bgIndex + lighten,
        // always lighten on hover no matter the scheme
        backgroundPress: bgIndex + darken,
        // always darken on press no matter the theme
        backgroundFocus: bgIndex + darken,
        borderColor,
        borderColorHover: borderColor + lighten,
        borderColorPress: borderColor + darken,
        borderColorFocus: borderColor,
        color: -bgIndex,
        colorHover: -bgIndex - 1,
        colorPress: -bgIndex,
        colorFocus: -bgIndex - 1,
        colorTransparent: -1,
        placeholderColor: -bgIndex - 3,
        outlineColor: -2
      },
      surface1 = {
        background: base.background + 1,
        backgroundHover: base.backgroundHover + 1,
        backgroundPress: base.backgroundPress + 1,
        backgroundFocus: base.backgroundFocus + 1,
        borderColor: base.borderColor + 1,
        borderColorHover: base.borderColorHover + 1,
        borderColorFocus: base.borderColorFocus + 1,
        borderColorPress: base.borderColorPress + 1
      },
      surface2 = {
        background: base.background + 2,
        backgroundHover: base.backgroundHover + 2,
        backgroundPress: base.backgroundPress + 2,
        backgroundFocus: base.backgroundFocus + 2,
        borderColor: base.borderColor + 2,
        borderColorHover: base.borderColorHover + 2,
        borderColorFocus: base.borderColorFocus + 2,
        borderColorPress: base.borderColorPress + 2
      },
      surface3 = {
        background: base.background + 3,
        backgroundHover: base.backgroundHover + 3,
        backgroundPress: base.backgroundPress + 3,
        backgroundFocus: base.backgroundFocus + 3,
        borderColor: base.borderColor + 3,
        borderColorHover: base.borderColorHover + 3,
        borderColorFocus: base.borderColorFocus + 3,
        borderColorPress: base.borderColorPress + 3
      },
      alt1 = {
        color: base.color - 1,
        colorHover: base.colorHover - 1,
        colorPress: base.colorPress - 1,
        colorFocus: base.colorFocus - 1
      },
      alt2 = {
        color: base.color - 2,
        colorHover: base.colorHover - 2,
        colorPress: base.colorPress - 2,
        colorFocus: base.colorFocus - 2
      },
      inverse = Object.fromEntries(Object.entries(base).map(([key, index]) => [key, -index]));
    return {
      base,
      surface1,
      surface2,
      surface3,
      alt1,
      alt2,
      inverse
    };
  },
  defaultTemplates = getTemplates();
export { defaultTemplates };
