// @tamagui/static/dist/extractor/loadTamagui.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf, __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: !0 }) : target,
  mod
)), __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var loadTamagui_exports = {};
__export(loadTamagui_exports, {
  esbuildWatchFiles: () => esbuildWatchFiles,
  generateThemesAndLog: () => generateThemesAndLog,
  getOptions: () => getOptions,
  loadTamagui: () => loadTamagui,
  loadTamaguiBuildConfigSync: () => loadTamaguiBuildConfigSync,
  loadTamaguiSync: () => loadTamaguiSync,
  resolveWebOrNativeSpecificEntry: () => resolveWebOrNativeSpecificEntry
});
module.exports = __toCommonJS(loadTamagui_exports);
var import_node_path = require("node:path"), import_cli_color = require("@tamagui/cli-color"), import_esbuild = __toESM(require("esbuild")), fsExtra = __toESM(require("fs-extra")), import_constants = require("../constants"), import_requireTamaguiCore = require("../helpers/requireTamaguiCore"), import_registerRequire = require("../registerRequire"), import_bundleConfig = require("./bundleConfig"), import_getTamaguiConfigPathFromOptionsConfig = require("./getTamaguiConfigPathFromOptionsConfig"), import_regenerateConfig = require("./regenerateConfig");
const getFilledOptions = (propsIn) => ({
  // defaults
  platform: "web",
  config: "tamagui.config.ts",
  components: ["tamagui"],
  ...propsIn
});
let isLoadingPromise;
async function loadTamagui(propsIn) {
  if (isLoadingPromise) return await isLoadingPromise;
  let resolvePromise, rejectPromise;
  isLoadingPromise = new Promise((res, rej) => {
    resolvePromise = res, rejectPromise = rej;
  });
  try {
    const props = getFilledOptions(propsIn), bundleInfo = await (0, import_bundleConfig.getBundledConfig)(props);
    return bundleInfo ? (await generateThemesAndLog(props), (0, import_bundleConfig.hasBundledConfigChanged)() ? (process.env.NODE_ENV === "development" && await (0, import_regenerateConfig.regenerateConfig)(props, bundleInfo), resolvePromise(bundleInfo), bundleInfo) : (resolvePromise(bundleInfo), bundleInfo)) : (console.warn(
      "No bundled config generated, maybe an error in bundling. Set DEBUG=tamagui and re-run to get logs."
    ), resolvePromise(null), null);
  } catch (err) {
    throw rejectPromise(), err;
  } finally {
    isLoadingPromise = null;
  }
}
let waiting = !1;
const generateThemesAndLog = async (options, force = !1) => {
  if (!waiting && options.themeBuilder)
    try {
      if (waiting = !0, await new Promise((res) => setTimeout(res, 30)), await (0, import_regenerateConfig.generateTamaguiThemes)(options, force) && ((0, import_cli_color.colorLog)(
        import_cli_color.Color.FgYellow,
        `    \u27A1 [tamagui] generated themes: ${(0, import_node_path.relative)(
          process.cwd(),
          options.themeBuilder.output
        )}`
      ), options.outputCSS)) {
        const loadedConfig = (0, import_bundleConfig.getLoadedConfig)();
        loadedConfig && await (0, import_bundleConfig.writeTamaguiCSS)(options.outputCSS, loadedConfig);
      }
    } finally {
      waiting = !1;
    }
}, last = {}, lastVersion = {};
function loadTamaguiBuildConfigSync(tamaguiOptions) {
  const buildFilePath = tamaguiOptions?.buildFile ?? "tamagui.build.ts";
  if (fsExtra.existsSync(buildFilePath)) {
    const registered = (0, import_registerRequire.registerRequire)("web");
    try {
      const out = require(buildFilePath).default;
      if (!out)
        throw new Error(`No default export found in ${buildFilePath}: ${out}`);
      tamaguiOptions = {
        ...tamaguiOptions,
        ...out
      };
    } finally {
      registered.unregister();
    }
  }
  if (!tamaguiOptions)
    throw new Error(
      "No tamagui build options found either via input props or at tamagui.build.ts"
    );
  return {
    config: "tamagui.config.ts",
    components: ["@tamagui/core"],
    ...tamaguiOptions
  };
}
function loadTamaguiSync({
  forceExports,
  cacheKey,
  ...propsIn
}) {
  const key = JSON.stringify(propsIn);
  if (last[key] && !(0, import_bundleConfig.hasBundledConfigChanged)() && (!lastVersion[key] || lastVersion[key] === cacheKey))
    return last[key];
  lastVersion[key] = cacheKey || "";
  const props = getFilledOptions(propsIn);
  process.env.IS_STATIC = "is_static", process.env.TAMAGUI_IS_SERVER = "true";
  const { unregister } = (0, import_registerRequire.registerRequire)(props.platform || "web", {
    proxyWormImports: !!forceExports
  });
  try {
    const devValueOG = globalThis.__DEV__;
    globalThis.__DEV__ = process.env.NODE_ENV === "development";
    try {
      let tamaguiConfig = null;
      if (propsIn.config) {
        const configPath = (0, import_getTamaguiConfigPathFromOptionsConfig.getTamaguiConfigPathFromOptionsConfig)(propsIn.config), exp = require(configPath);
        if (tamaguiConfig = exp.default || exp.config || exp, !tamaguiConfig || !tamaguiConfig.parsed) {
          const confPath = require.resolve(configPath);
          throw new Error(`Can't find valid config in ${confPath}:
          
  Be sure you "export default" or "export const config" the config.`);
        }
        if (tamaguiConfig) {
          const { createTamagui } = (0, import_requireTamaguiCore.requireTamaguiCore)(props.platform || "web");
          createTamagui(tamaguiConfig);
        }
      }
      const components = (0, import_bundleConfig.loadComponents)(props, forceExports);
      if (!components)
        throw new Error("No components loaded");
      process.env.DEBUG === "tamagui" && console.info("components", components), process.env.IS_STATIC = void 0, globalThis.__DEV__ = devValueOG;
      const info = {
        components,
        tamaguiConfig,
        nameToPaths: (0, import_registerRequire.getNameToPaths)()
      };
      if (tamaguiConfig) {
        const { outputCSS } = props;
        outputCSS && (0, import_bundleConfig.writeTamaguiCSS)(outputCSS, tamaguiConfig), (0, import_regenerateConfig.regenerateConfigSync)(props, info);
      }
      return last[key] = {
        ...info,
        cached: !0
      }, info;
    } catch (err) {
      err instanceof Error ? !import_constants.SHOULD_DEBUG && !forceExports ? (console.warn(
        "Error loading tamagui.config.ts (set DEBUG=tamagui to see full stack), running tamagui without custom config"
      ), console.info(`

    ${err.message}

`)) : import_constants.SHOULD_DEBUG && console.error(err) : console.error("Error loading tamagui.config.ts", err);
      const { createTamagui } = (0, import_requireTamaguiCore.requireTamaguiCore)(props.platform || "web"), { getDefaultTamaguiConfig } = require("@tamagui/config-default");
      return {
        components: [],
        tamaguiConfig: createTamagui(getDefaultTamaguiConfig()),
        nameToPaths: {}
      };
    }
  } finally {
    unregister();
  }
}
async function getOptions({
  root = process.cwd(),
  tsconfigPath = "tsconfig.json",
  tamaguiOptions,
  host,
  debug
} = {}) {
  const dotDir = (0, import_node_path.join)(root, ".tamagui");
  let pkgJson = {};
  try {
    pkgJson = await fsExtra.readJSON((0, import_node_path.join)(root, "package.json"));
  } catch {
  }
  return {
    mode: process.env.NODE_ENV === "production" ? "production" : "development",
    root,
    host: host || "127.0.0.1",
    pkgJson,
    debug,
    tsconfigPath,
    tamaguiOptions: {
      platform: "web",
      components: ["tamagui"],
      ...tamaguiOptions,
      config: tamaguiOptions?.config ?? await getDefaultTamaguiConfigPath(root, tamaguiOptions?.config)
    },
    paths: {
      root,
      dotDir,
      conf: (0, import_node_path.join)(dotDir, "tamagui.config.json"),
      types: (0, import_node_path.join)(dotDir, "types.json")
    }
  };
}
function resolveWebOrNativeSpecificEntry(entry) {
  const workspaceRoot = (0, import_node_path.resolve)(), resolved = require.resolve(entry, { paths: [workspaceRoot] }), ext = (0, import_node_path.extname)(resolved), fileName = (0, import_node_path.basename)(resolved).replace(ext, ""), specificFile = (0, import_node_path.join)((0, import_node_path.dirname)(resolved), fileName + "." + "web" + ext);
  return fsExtra.existsSync(specificFile) ? specificFile : entry;
}
const defaultPaths = ["tamagui.config.ts", (0, import_node_path.join)("src", "tamagui.config.ts")];
let hasWarnedOnce = !1;
async function getDefaultTamaguiConfigPath(root, configPath) {
  const searchPaths = [
    ...new Set(
      [configPath, ...defaultPaths].filter(Boolean).map((p) => (0, import_node_path.join)(root, p))
    )
  ];
  for (const path of searchPaths)
    if (await fsExtra.pathExists(path))
      return path;
  hasWarnedOnce || (hasWarnedOnce = !0, console.warn(`Warning: couldn't find tamagui.config.ts in the following paths given configuration "${configPath}":
    ${searchPaths.join(`
  `)}
  `));
}
async function esbuildWatchFiles(entry, onChanged) {
  let hasRunOnce = !1;
  const context = await import_esbuild.default.context({
    bundle: !0,
    entryPoints: [entry],
    resolveExtensions: [".ts", ".tsx", ".js", ".mjs"],
    logLevel: "silent",
    write: !1,
    alias: {
      "@react-native/normalize-color": "@tamagui/proxy-worm",
      "react-native-web": "@tamagui/proxy-worm",
      "react-native": "@tamagui/proxy-worm"
    },
    plugins: [
      // to log what its watching:
      // {
      //   name: 'test',
      //   setup({ onResolve }) {
      //     onResolve({ filter: /.*/ }, (args) => {
      //       console.log('wtf', args.path)
      //     })
      //   },
      // },
      {
        name: "on-rebuild",
        setup({ onEnd, onResolve }) {
          onResolve({ filter: /^[^.\/]|^\.[^.\/]|^\.\.[^\/]/ }, (args) => ({ path: args.path, external: !0 })), onEnd(() => {
            hasRunOnce ? onChanged() : hasRunOnce = !0;
          });
        }
      }
    ]
  });
  return context.watch(), () => {
    context.dispose();
  };
}
