// @tamagui/config/dist/esm/v4.js
import { shorthands } from "@tamagui/shorthands/v2";
import { tokens } from "@tamagui/themes/v4";
import { animations } from "./v3-animations";
import { fonts } from "./v4-fonts";
import { media, mediaQueryDefaultActive } from "./v4-media";
import { shorthands as shorthands2 } from "@tamagui/shorthands/v3";
import { tokens as tokens2 } from "@tamagui/themes/v3-themes";
import { fonts as fonts2 } from "./fonts";
import { media as media2, mediaQueryDefaultActive as mediaQueryDefaultActive2 } from "./media";
import { animations as animations2 } from "./v4-animations";
const config = {
  animations,
  media,
  shorthands,
  tokens,
  fonts,
  selectionStyles: (theme) => theme.color5 ? {
    backgroundColor: theme.color5,
    color: theme.color11
  } : null,
  settings: {
    mediaQueryDefaultActive,
    defaultFont: "body",
    fastSchemeChange: !0,
    shouldAddPrefersColorThemes: !0,
    allowedStyleValues: "somewhat-strict-web",
    themeClassNameOnRoot: !0
  }
};
export {
  animations2 as animations,
  config,
  fonts2 as fonts,
  media2 as media,
  mediaQueryDefaultActive2 as mediaQueryDefaultActive,
  shorthands2 as shorthands,
  tokens2 as tokens
};
