// @tamagui/config/dist/esm/v4.mjs
import { shorthands } from "@tamagui/shorthands/v2";
import { tokens } from "@tamagui/themes/v4";
import { animations } from "./v3-animations.mjs";
import { fonts } from "./v4-fonts.mjs";
import { media, mediaQueryDefaultActive } from "./v4-media.mjs";
import { shorthands as shorthands2 } from "@tamagui/shorthands/v3";
import { tokens as tokens2 } from "@tamagui/themes/v3-themes";
import { fonts as fonts2 } from "./fonts.mjs";
import { media as media2, mediaQueryDefaultActive as mediaQueryDefaultActive2 } from "./media.mjs";
import { animations as animations2 } from "./v4-animations.mjs";
const config = {
  animations,
  media,
  shorthands,
  tokens,
  fonts,
  selectionStyles: theme => theme.color5 ? {
    backgroundColor: theme.color5,
    color: theme.color11
  } : null,
  settings: {
    mediaQueryDefaultActive,
    defaultFont: "body",
    fastSchemeChange: !0,
    shouldAddPrefersColorThemes: !0,
    allowedStyleValues: "somewhat-strict-web",
    themeClassNameOnRoot: !0
  }
};
export { animations2 as animations, config, fonts2 as fonts, media2 as media, mediaQueryDefaultActive2 as mediaQueryDefaultActive, shorthands2 as shorthands, tokens2 as tokens };
