// @tamagui/react-native-web-internals/dist/esm/modules/InteractionManager.native.js
import requestIdleCallback, { cancelIdleCallback } from "./requestIdleCallback";
var InteractionManager = {
  Events: {
    interactionStart: "interactionStart",
    interactionComplete: "interactionComplete"
  },
  /**
  * Schedule a function to run after all interactions have completed.
  */
  runAfterInteractions(task) {
    var handle, promise = new Promise(function(resolve) {
      handle = requestIdleCallback(function() {
        task ? resolve(task()) : resolve();
      });
    });
    return {
      then: promise.then.bind(promise),
      done: promise.then.bind(promise),
      cancel: function() {
        cancelIdleCallback(handle);
      }
    };
  },
  /**
  * Notify manager that an interaction has started.
  */
  createInteractionHandle() {
    return 1;
  },
  /**
  * Notify manager that an interaction has completed.
  */
  clearInteractionHandle(handle) {
  },
  addListener: function() {
  }
}, InteractionManager_default = InteractionManager;
export {
  InteractionManager,
  InteractionManager_default as default
};
