// @tamagui/react-native-web-lite/dist/cjs/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf,
  __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
    value: mod,
    enumerable: !0
  }) : target, mod)),
  __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
    value: !0
  }), mod);
var index_exports = {};
__export(index_exports, {
  AccessibilityInfo: () => import_AccessibilityInfo.default,
  AccessibilityUtil: () => import_react_native_web_internals.AccessibilityUtil,
  ActivityIndicator: () => import_ActivityIndicator.default,
  Alert: () => import_Alert.default,
  Animated: () => import_Animated.default,
  AppRegistry: () => import_AppRegistry.default,
  AppState: () => import_AppState.default,
  Appearance: () => import_Appearance.default,
  BackHandler: () => import_BackHandler.default,
  Clipboard: () => import_Clipboard.default,
  DeviceEmitter: () => import_DeviceEmitter.default,
  DeviceEventEmitter: () => import_DeviceEmitter2.default,
  DeviceInfo: () => import_DeviceInfo.default,
  Dimensions: () => import_Dimensions.default,
  DrawerLayoutAndroid: () => import_UnimplementedView.default,
  Easing: () => import_Easing.default,
  FlatList: () => import_FlatList.default,
  I18nManager: () => import_I18nManager.default,
  Image: () => import_Image.default,
  ImageBackground: () => import_ImageBackground.default,
  ImageLoader: () => import_react_native_web_internals.ImageLoader,
  InteractionManager: () => import_react_native_web_internals.InteractionManager,
  Keyboard: () => import_Keyboard.default,
  KeyboardAvoidingView: () => import_KeyboardAvoidingView.default,
  Linking: () => import_Linking.default,
  LocaleProvider: () => import_react_native_web_internals.LocaleProvider,
  LogBox: () => import_LogBox.default,
  Modal: () => import_Modal.default,
  NativeEventEmitter: () => import_NativeEventEmitter.default,
  NativeModules: () => import_NativeModules.default,
  PanResponder: () => import_PanResponder.default,
  PixelRatio: () => import_PixelRatio.default,
  Platform: () => import_react_native_web_internals.Platform,
  Pressable: () => import_Pressable.default,
  RefreshControl: () => import_RefreshControl.default,
  SafeAreaView: () => import_SafeAreaView.default,
  ScrollView: () => import_ScrollView.default,
  SectionList: () => import_UnimplementedView5.default,
  Share: () => import_Share.default,
  StatusBar: () => import_StatusBar.default,
  StyleSheet: () => import_react_native_web_internals.StyleSheet,
  Switch: () => import_UnimplementedView2.default,
  Text: () => import_Text.default,
  TextAncestorContext: () => import_react_native_web_internals.TextAncestorContext,
  TextInput: () => import_TextInput.default,
  Touchable: () => import_TouchableOpacity.default,
  TouchableHighlight: () => import_UnimplementedView3.default,
  TouchableNativeFeedback: () => import_UnimplementedView4.default,
  TouchableOpacity: () => import_TouchableOpacity2.default,
  TouchableWithoutFeedback: () => import_TouchableWithoutFeedback.default,
  UIManager: () => import_react_native_web_internals.UIManager,
  Vibration: () => import_Vibration.default,
  View: () => import_View.default,
  VirtualizedList: () => import_VirtualizedList.default,
  canUseDOM: () => import_react_native_web_internals.canUseDOM,
  clickProps: () => import_react_native_web_internals.clickProps,
  createDOMProps: () => import_react_native_web_internals.createDOMProps,
  dismissKeyboard: () => import_react_native_web_internals.dismissKeyboard,
  findNodeHandle: () => import_findNodeHandle.findNodeHandle,
  flattenStyle: () => import_react_native_web_internals.flattenStyle,
  isWebColor: () => import_react_native_web_internals.isWebColor,
  mergeRefs: () => import_react_native_web_internals.mergeRefs,
  normalizeColor: () => import_react_native_web_internals.normalizeColor,
  processColor: () => import_react_native_web_internals.processColor,
  processStyle: () => import_react_native_web_internals.processStyle,
  render: () => import_render.default,
  requireNativeComponent: () => requireNativeComponent,
  unmountComponentAtNode: () => import_react_dom.unmountComponentAtNode,
  unstable_createElement: () => import_createElement.default,
  useColorScheme: () => import_useColorScheme.default,
  useEvent: () => import_react_native_web_internals.useEvent,
  useHover: () => import_react_native_web_internals.useHover,
  useLayoutEffect: () => import_react_native_web_internals.useLayoutEffect,
  useLocaleContext: () => import_useLocaleContext.useLocaleContext,
  useMergeRefs: () => import_react_native_web_internals.useMergeRefs,
  usePlatformMethods: () => import_react_native_web_internals.usePlatformMethods,
  useWindowDimensions: () => import_useWindowDimensions.default
});
module.exports = __toCommonJS(index_exports);
var import_createElement = __toESM(require("./createElement/index.cjs")),
  import_react_native_web_internals = require("@tamagui/react-native-web-internals"),
  import_render = __toESM(require("./render/index.cjs")),
  import_react_dom = require("react-dom"),
  import_NativeModules = __toESM(require("./NativeModules/index.cjs")),
  import_findNodeHandle = require("./findNodeHandle.cjs"),
  import_Easing = __toESM(require("./vendor/react-native/Animated/Easing.cjs")),
  import_Animated = __toESM(require("./vendor/react-native/Animated/Animated.cjs")),
  import_NativeEventEmitter = __toESM(require("./vendor/react-native/NativeEventEmitter/index.cjs")),
  import_AccessibilityInfo = __toESM(require("./AccessibilityInfo/index.cjs")),
  import_Alert = __toESM(require("./Alert/index.cjs")),
  import_Appearance = __toESM(require("./Appearance/index.cjs")),
  import_AppRegistry = __toESM(require("./AppRegistry/index.cjs")),
  import_AppState = __toESM(require("./AppState/index.cjs")),
  import_BackHandler = __toESM(require("./BackHandler/index.cjs")),
  import_Clipboard = __toESM(require("./Clipboard/index.cjs")),
  import_DeviceInfo = __toESM(require("./DeviceInfo/index.cjs")),
  import_DeviceEmitter = __toESM(require("./DeviceEmitter.cjs")),
  import_DeviceEmitter2 = __toESM(require("./DeviceEmitter.cjs")),
  import_Dimensions = __toESM(require("./Dimensions/index.cjs")),
  import_I18nManager = __toESM(require("./I18nManager/index.cjs")),
  import_Keyboard = __toESM(require("./Keyboard/index.cjs")),
  import_Linking = __toESM(require("./Linking/index.cjs")),
  import_PanResponder = __toESM(require("./PanResponder/index.cjs")),
  import_PixelRatio = __toESM(require("./PixelRatio/index.cjs")),
  import_Share = __toESM(require("./Share/index.cjs")),
  import_Vibration = __toESM(require("./Vibration/index.cjs")),
  import_UnimplementedView = __toESM(require("./UnimplementedView.cjs")),
  import_UnimplementedView2 = __toESM(require("./UnimplementedView.cjs")),
  import_VirtualizedList = __toESM(require("./vendor/react-native/VirtualizedList/index.cjs")),
  import_FlatList = __toESM(require("./FlatList.cjs")),
  import_UnimplementedView3 = __toESM(require("./UnimplementedView.cjs")),
  import_UnimplementedView4 = __toESM(require("./UnimplementedView.cjs")),
  import_UnimplementedView5 = __toESM(require("./UnimplementedView.cjs")),
  import_TouchableOpacity = __toESM(require("./TouchableOpacity.cjs")),
  import_TouchableOpacity2 = __toESM(require("./TouchableOpacity.cjs")),
  import_TouchableWithoutFeedback = __toESM(require("./TouchableWithoutFeedback.cjs")),
  import_ActivityIndicator = __toESM(require("./ActivityIndicator/index.cjs")),
  import_Image = __toESM(require("./Image/index.cjs")),
  import_ImageBackground = __toESM(require("./ImageBackground/index.cjs")),
  import_KeyboardAvoidingView = __toESM(require("./KeyboardAvoidingView/index.cjs")),
  import_Modal = __toESM(require("./Modal/index.cjs")),
  import_Pressable = __toESM(require("./Pressable/index.cjs")),
  import_RefreshControl = __toESM(require("./RefreshControl/index.cjs")),
  import_SafeAreaView = __toESM(require("./SafeAreaView/index.cjs")),
  import_ScrollView = __toESM(require("./ScrollView/index.cjs")),
  import_StatusBar = __toESM(require("./StatusBar/index.cjs")),
  import_Text = __toESM(require("./Text/index.cjs")),
  import_TextInput = __toESM(require("./TextInput/index.cjs")),
  import_View = __toESM(require("./View/index.cjs")),
  import_LogBox = __toESM(require("./LogBox/index.cjs")),
  import_useColorScheme = __toESM(require("./useColorScheme/index.cjs")),
  import_useLocaleContext = require("./useLocaleContext/index.cjs"),
  import_useWindowDimensions = __toESM(require("./useWindowDimensions/index.cjs"));
function requireNativeComponent(name) {
  return function () {
    return null;
  };
}
