// @tamagui/config/dist/cjs/v4.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var v4_exports = {};
__export(v4_exports, {
  animations: () => import_v4_animations.animations,
  config: () => config,
  fonts: () => import_fonts.fonts,
  media: () => import_media.media,
  mediaQueryDefaultActive: () => import_media.mediaQueryDefaultActive,
  shorthands: () => import_v3.shorthands,
  tokens: () => import_v3_themes.tokens
});
module.exports = __toCommonJS(v4_exports);
var import_v2 = require("@tamagui/shorthands/v2"), import_v4 = require("@tamagui/themes/v4"), import_v3_animations = require("./v3-animations"), import_v4_fonts = require("./v4-fonts"), import_v4_media = require("./v4-media"), import_v3 = require("@tamagui/shorthands/v3"), import_v3_themes = require("@tamagui/themes/v3-themes"), import_fonts = require("./fonts"), import_media = require("./media"), import_v4_animations = require("./v4-animations");
const config = {
  animations: import_v3_animations.animations,
  media: import_v4_media.media,
  shorthands: import_v2.shorthands,
  tokens: import_v4.tokens,
  fonts: import_v4_fonts.fonts,
  selectionStyles: (theme) => theme.color5 ? {
    backgroundColor: theme.color5,
    color: theme.color11
  } : null,
  settings: {
    mediaQueryDefaultActive: import_v4_media.mediaQueryDefaultActive,
    defaultFont: "body",
    fastSchemeChange: !0,
    shouldAddPrefersColorThemes: !0,
    allowedStyleValues: "somewhat-strict-web",
    themeClassNameOnRoot: !0
  }
};
